pacsman/filesystem_dev_client.py

```diff
@@ -35,7 +40,29 @@
 from .base_client import BaseDicomClient, PRIVATE_ID
 from .utils import process_and_write_png, copy_dicom_attributes, dicom_filename
 
-logger = logging.getLogger(__name__)
+
+INDEX_FILENAME = ".pacsman_index"
+
+
+def _default_empty_list():
+    """
+    This replaces `lambda: []` for defaultdicts, as pickle does not serialize lambdas.
+    """
+    return []
+
+
+class FilesystemDicomIndex():
+    def __init__(self):
+        self.dicom_source_dir_hash: str = ''
+        self.filepaths: List[str] = []
+        self.patient_id_to_filepaths: Dict[str, List[int]] = \
+            defaultdict(_default_empty_list)
+        self.series_id_to_filepaths: Dict[str, List[int]] = \
+            defaultdict(_default_empty_list)
+        self.study_id_to_filepaths: Dict[str, List[int]] = \
+            defaultdict(_default_empty_list)
+        self.patient_name_to_filepaths: Dict[str, List[int]] = \
+            defaultdict(_default_empty_list)
 
 
 class FilesystemDicomClient(BaseDicomClient):
```

Comment:
Just a thought, but would passing in the index filename into the `FilesystemDicomClient`, vs using a global const, make it easier to test?

Reply:
It would make it a little easier to test, but each user when then have to concern themselves with the index file location and name, as opposed to it being handled in the background. There are definitely use cases for this, like putting the index and the data on separate volumes, but they might be beyond the scope of the filesystem client at that point.